File: Tools/setuptool/HemeLbSetupTool/Model/test_TriangleSorter.py
import numpy as np
from TestResources.sphere import GetSphereNumpy
from . import TriangleSorter
from .HemeOct import Tree
from nose.tools import set_trace
def trees_with_triIds_equal(t1, t2, tri_level):
    if t1 != t2:
        return False
    # If here, they're structurally the same, check triIds
    for n1 in t1.IterDepthFirst(tri_level, tri_level):
        n2 = t2.GetNode(n1.levels, n1.offset)
        if not np.all(n1.triIds == n2.triIds):
            return False
        continue
    # None of the nodes differ so we are true!
    return True
        
def mk_trivial():
    # Put the tri points on a square, in the (0,0,0) octant
    points = np.array([(1.2, 1.2, 1.2),
                       (1.2, 1.2, 2.2),
                       (1.2, 2.2, 1.2),
                       (1.2, 2.2, 2.2)], dtype=float)
    # Define the tris
    triangles = np.array([(0,1,2),
                          (2,1,3)], dtype=int)
    normals = np.array([[-1, 0, 0],
                        [-1, 0, 0]], dtype=float)
    return points, triangles, normals

def mk_trivial2():

    points = np.array([(1.2, 1.2, 1.2),
                       (1.2, 1.2, 2.2),
                       (1.2, 2.2, 1.2),
                       (1.2, 2.2, 2.2),
                       (1.2, 1.2, 3.2),
                       (1.2, 2.2, 3.2)], dtype=float)

    triangles = np.array([(0,1,2),
                          (2,1,3),(0,4,2)], dtype=int)
    normals = np.array([[-1, 0, 0],
                        [-1, 0, 0],[-1, 0, 0]], dtype=float)

    return points, triangles, normals

def test_trivial():
    # 16 cube
    levels = 4
    # put triangles onto the 8 cube level
    tri_level = 3
    points, triangles, normals = mk_trivial()
    
    # Sort!
    tree = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    
    assert tree.levels == levels
    i = tri_level
    for node in tree.IterDepthFirst():
        assert np.all(node.offset == 0)
        assert node.levels == i
        i += 1
        continue
    assert i == levels +1
    
    node = tree.GetNode(tri_level, np.array((0,0,0)))
    ids = node.triIds
    assert np.all(ids == (0,1))

def test_trivial_deep():
    # 1,024 cube
    levels = 18
    # put triangles onto the 8 cube level
    tri_level = 17
    points, triangles, normals = mk_trivial()
    
    # Sort!
    tree = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    
    assert tree.levels == levels
    i = tri_level
    for node in tree.IterDepthFirst():
        assert np.all(node.offset == 0)
        assert node.levels == i
        i += 1
        continue
    assert i == levels +1
    
    node = tree.GetNode(tri_level, np.array((0,0,0)))
    ids = node.triIds
    assert np.all(ids == (0,1))

def overlap1d(amin, amax, bmin, bmax):
    return np.logical_and(amin < bmax, amax > bmin)
def overlap3d(amin, amax, bmin, bmax):
    return np.logical_and(
        np.logical_and(
            overlap1d(amin[..., 0], amax[..., 0], bmin[..., 0], bmax[..., 0]),
            overlap1d(amin[..., 1], amax[..., 1], bmin[..., 1], bmax[..., 1])),
        overlap1d(amin[..., 2], amax[..., 2], bmin[..., 2], bmax[..., 2]))
    
def test_sphere():
    levels = 5
    tri_level = 3
    points, triangles, normals = GetSphereNumpy()
    tree = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    
    seen_tris = set()
    for node in tree.IterDepthFirst(tri_level, tri_level):
        n = len(node.triIds)
        node_points = np.zeros((n, 3, 3))
        node_point_ids = triangles[node.triIds]
        for iPt in xrange(3):
            node_points[:,iPt,:] = points[node_point_ids[:,iPt]]
        
        tri_min = node_points.min(axis=1)
        tri_max = node_points.max(axis=1)
        
        nd_min = (node.offset - 1)[np.newaxis,:]
        nd_max = (node.offset + 2**tri_level + 1)[np.newaxis,:]
        
        assert np.all(overlap3d(tri_min, tri_max, nd_min, nd_max))
            
        seen_tris.update(node.triIds)
    allids = sorted(seen_tris)
    assert np.all(np.equal(allids, np.arange(len(triangles))))
    
def test_merge_identity():
    # 16 cube
    levels = 4
    # put triangles onto the 8 cube level
    tri_level = 3
    points, triangles, normals = mk_trivial()
    
    tree = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    nulltree = Tree(levels)
    
    s = TriangleSorter.TreeSummer(levels, tri_level)
    s.add(tree)
    s.add(nulltree)
    new = s.finish()
    
    assert trees_with_triIds_equal(new, tree, tri_level)

def test_merge_simple():
    # 16 cube
    levels = 4
    # put triangles onto the 8 cube level
    tri_level = 3
    points, triangles, normals = mk_trivial()
    
    t1 = TriangleSorter.TrianglesToTree_Worker(levels, tri_level, points, triangles[:1], 0)
    t2 = TriangleSorter.TrianglesToTree_Worker(levels, tri_level, points, triangles[1:], 1)
    
    s = TriangleSorter.TreeSummer(levels, tri_level)
    s.add(t1)
    s.add(t2)
    new = s.finish()
    
    tree = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    assert trees_with_triIds_equal(new, tree, tri_level)

def test_merge_self():
    # Actually, t1 + t1 gives an invalid tree - we expect each triangle to be 
    # handled by exactly one process so adding support for ignoring duplicates 
    # is wasteful.
    
    # The final tree should therefore differ from t1
    
    # 16 cube
    levels = 4
    # put triangles onto the 8 cube level
    tri_level = 3
    points, triangles, normals = mk_trivial()
    
    t1 = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    
    s = TriangleSorter.TreeSummer(levels, tri_level)
    s.add(t1)
    s.add(t1)
    tree = s.finish()
    
    # Trees must be structurally identical
    assert tree == t1
    # Trees must have different IDs
    assert not trees_with_triIds_equal(tree, t1, tri_level)
    
def test_parallel():
    levels = 5
    tri_level = 3
    points, triangles, normals = GetSphereNumpy()
    serial = TriangleSorter.TrianglesToTree(levels, tri_level, points, triangles)
    
    p1 = TriangleSorter.TrianglesToTreeParallel(levels, tri_level, points, triangles, 1)
    p2 = TriangleSorter.TrianglesToTreeParallel(levels, tri_level, points, triangles, 2)
    p3 = TriangleSorter.TrianglesToTreeParallel(levels, tri_level, points, triangles, 3)
    p4 = TriangleSorter.TrianglesToTreeParallel(levels, tri_level, points, triangles, 4)
    p8 = TriangleSorter.TrianglesToTreeParallel(levels, tri_level, points, triangles, 8)
    assert trees_with_triIds_equal(serial, p1, tri_level)
    assert trees_with_triIds_equal(serial, p2, tri_level)
    assert trees_with_triIds_equal(serial, p3, tri_level)
    assert trees_with_triIds_equal(serial, p4, tri_level)
    assert trees_with_triIds_equal(serial, p8, tri_level)
    
    return

File: geometry-tool/HlbGmyTool/Model/OctGeneration/test_TriangleSorter.cpp
// This file is part of HemeLB and is Copyright (C)
// the HemeLB team and/or their institutions, as detailed in the
// file AUTHORS. This software is provided under the terms of the
// license in the file LICENSE.

#include <catch2/catch.hpp>
#include <set>

#include "TestResources/Meshes.h"
#include "TriangleSorter.h"

namespace hemelb::gmytool::oct {

struct TriangleSorterTests {
  void Trivial() {
    // 16 cube
    TriTree::Int levels = 4;
    // put triangles onto the 8 cube level
    TriTree::Int tri_level = 3;
    auto triv = SimpleMeshFactory::MkTrivial();
    // Sort!
    auto tree =
        TrianglesToTreeSerial(levels, tri_level, triv->points, triv->triangles);

    REQUIRE(tree.Level() == levels);
    TriTree::Int i = tri_level;
    tree.IterDepthFirst([&i](TriTree::NodePtr node) mutable {
      REQUIRE(node->X() == 0);
      REQUIRE(node->Y() == 0);
      REQUIRE(node->Z() == 0);
      REQUIRE(node->Level() == i);
      ++i;
    });

    auto node = tree.Get(0, 0, 0, tri_level);
    auto ids = node->Data();
    std::sort(ids.begin(), ids.end());
    REQUIRE(ids.size() == 2);
    REQUIRE(ids.find(0) != ids.end());
    REQUIRE(ids.find(1) != ids.end());
  }

  void TrivialDeep() {
    // 1024 cube
    TriTree::Int levels = 10;
    // put triangles onto the 8 cube level
    TriTree::Int tri_level = 3;
    auto triv = SimpleMeshFactory::MkTrivial();
    // Sort!
    auto tree =
        TrianglesToTreeSerial(levels, tri_level, triv->points, triv->triangles);

    REQUIRE(tree.Level() == levels);
    TriTree::Int i = tri_level;
    tree.IterDepthFirst([&i](TriTree::NodePtr node) mutable {
      REQUIRE(node->X() == 0);
      REQUIRE(node->Y() == 0);
      REQUIRE(node->Z() == 0);
      REQUIRE(node->Level() == i);
      ++i;
    });

    auto node = tree.Get(0, 0, 0, tri_level);
    auto ids = node->Data();
    std::sort(ids.begin(), ids.end());
    REQUIRE(ids.size() == 2);
    REQUIRE(ids.find(0) != ids.end());
    REQUIRE(ids.find(1) != ids.end());
  }

  template <class T>
  bool overlap1d(T aMin, T aMax, T bMin, T bMax) {
    return (aMin < bMax) && (aMax > bMin);
  }

  bool overlap3d(const Vector& aMin,
                 const Vector& aMax,
                 const Vector& bMin,
                 const Vector& bMax) {
    bool ans = true;
    for (auto d = 0; d < 3; ++d)
      ans &= overlap1d(aMin[d], aMax[d], bMin[d], bMax[d]);

    return ans;
  }

  void Sphere() {
    TriTree::Int levels = 5;
    TriTree::Int tri_level = 3;

    auto sphere = SimpleMeshFactory::MkSphere();
    auto tree = TrianglesToTreeSerial(levels, tri_level, sphere->points,
                                      sphere->triangles);

    std::set<int> seen_tris;

    tree.IterDepthFirst(tri_level, tri_level, [&](TriTree::NodePtr node) {
      const auto& triIds = node->Data();
      const Vector offset(node->X(), node->Y(), node->Z());
      auto node_min = offset - Vector{1};
      auto node_max = offset + Vector{(1 << tri_level) + 1};

      for (auto iTri : triIds) {
        Vector tri_min(std::numeric_limits<double>::infinity());
        Vector tri_max(0);
        for (auto iPt : sphere->triangles[iTri]) {
          tri_min.UpdatePointwiseMin(sphere->points[iPt]);
          tri_max.UpdatePointwiseMax(sphere->points[iPt]);
        }

        REQUIRE(overlap3d(tri_min, tri_max, node_min, node_max));
        seen_tris.insert(triIds.begin(), triIds.end());
      }
    });

    // Assert that we have seen all the triangles in the input
    REQUIRE(seen_tris.size() == sphere->triangles.size());
    int i = 0;
    for (auto seen_id : seen_tris) {
      REQUIRE(seen_id == i);
      ++i;
    }
  }

  bool trees_with_triIds_equal(TriTree& t1,
                               TriTree& t2,
                               TriTree::Int tri_level) {
    bool eq = true;
    try {
      t1.IterDepthFirst(tri_level, tri_level, [&](TriTree::NodePtr n1) {
        auto n2 = t2.Get(n1->X(), n1->Y(), n1->Z(), n1->Level());
        if (n1->Data() != n2->Data())
          throw 0;
      });
    } catch (int& i) {
      eq = false;
    }
    return eq;
  }
  void MergeIdentity() {
    // 16 cube
    TriTree::Int levels = 4;
    // put triangles onto the 8 cube level
    TriTree::Int tri_level = 3;
    auto triv = SimpleMeshFactory::MkTrivial();

    auto tree =
        TrianglesToTreeSerial(levels, tri_level, triv->points, triv->triangles);
    TriTree nulltree(levels);

    TreeSummer s(levels, tri_level);
    s.Add(tree);
    s.Add(nulltree);
    auto newtree = s.GetTree();
    REQUIRE(trees_with_triIds_equal(newtree, tree, tri_level));
  }

  void MergeSimple() {
    // 16
    TriTree::Int levels = 4;
    // put triangles onto the 8 cube level
    TriTree::Int tri_level = 3;
    auto triv = SimpleMeshFactory::MkTrivial();

    auto start = triv->triangles.cbegin();
    auto middle = start + 1;
    auto end = triv->triangles.cend();

    auto t1 = TrianglesToTree_Worker(levels, tri_level, triv->points, start,
                                     middle, 0);
    auto t2 =
        TrianglesToTree_Worker(levels, tri_level, triv->points, middle, end, 1);

    TreeSummer s(levels, tri_level);
    s.Add(t1);
    s.Add(t2);
    auto newtree = s.GetTree();

    auto tree =
        TrianglesToTreeSerial(levels, tri_level, triv->points, triv->triangles);
    REQUIRE(trees_with_triIds_equal(newtree, tree, tri_level));
  }

  void ParallelTrivial() {
    // 16 cube
    TriTree::Int levels = 4;
    // put triangles onto the 8 cube level
    TriTree::Int tri_level = 3;
    auto triv = SimpleMeshFactory::MkTrivial();
    // Sort!
    auto t1 =
        TrianglesToTreeSerial(levels, tri_level, triv->points, triv->triangles);
    auto t2 = TrianglesToTreeParallel(levels, tri_level, triv->points,
                                      triv->triangles, 2);
    REQUIRE(trees_with_triIds_equal(t1, t2, tri_level));
  }

  void ParallelSphere() {
    int np = GENERATE(0, 2, 3, 4);
    TriTree::Int levels = 5;
    TriTree::Int tri_level = 3;

    auto sphere = SimpleMeshFactory::MkSphere();
    auto t1 = TrianglesToTreeSerial(levels, tri_level, sphere->points,
                                    sphere->triangles);
    auto t2 = TrianglesToTreeParallel(levels, tri_level, sphere->points,
                                      sphere->triangles, np);
    REQUIRE(trees_with_triIds_equal(t1, t2, tri_level));
  }

  void Duct() {
    // 16 cube
    TriTree::Int levels = 4;
    TriTree::Int tri_level = 2;
    auto duct = SimpleMeshFactory::MkDuct();
    auto tree =
        TrianglesToTreeSerial(levels, tri_level, duct->points, duct->triangles);

    REQUIRE(tree.Get(0, 0, 0, 3));
    REQUIRE(tree.Get(0, 0, 8, 3));

    REQUIRE(tree.Get(0, 0, 0, 2));
    REQUIRE(tree.Get(0, 0, 4, 2));
    REQUIRE(tree.Get(0, 4, 0, 2));
    REQUIRE(tree.Get(0, 4, 4, 2));
    REQUIRE(tree.Get(4, 0, 0, 2));
    REQUIRE(tree.Get(4, 0, 4, 2));
    REQUIRE(tree.Get(4, 4, 0, 2));
    REQUIRE(tree.Get(4, 4, 4, 2));
  }
};

METHOD_AS_TEST_CASE(TriangleSorterTests::Trivial, "Trivial", "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::TrivialDeep,
                    "TrivialDeep",
                    "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::Sphere, "Sphere", "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::MergeIdentity,
                    "MergeIdentity",
                    "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::MergeSimple,
                    "MergeSimple",
                    "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::ParallelTrivial,
                    "ParallelTrivial",
                    "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::ParallelSphere,
                    "ParallelSphere",
                    "[triangle]");
METHOD_AS_TEST_CASE(TriangleSorterTests::Duct, "Duct", "[triangle]");

}  // namespace hemelb::gmytool::oct
